README.txt
Brief Installation Instructions (Binary Package)
================================================

This is rEFInd, an EFI boot manager. The binary package includes the
following files and subdirectories:

   File                             Description
   -----------------------------    -----------------------------
   refind/refind_ia32.efi           The main IA32 rEFInd binary
   refind/refind_x64.efi            The main x86-64 rEFInd binary
   refind/refind.conf-sample        A sample configuration file
   refind/icons/                    Subdirectory containing icons
   README.txt                       This file
   LICENSE.txt                      The original rEFIt license
   COPYING.txt                      The rEFInd license
   docs/                            Documentation in HTML format

To install the binary package, you must first access your EFI System
Partition (ESP). You can then place the files from the refind subdirectory
in a subdirectory of the ESP's EFI directory. You may omit the .efi binary
for the type of computer you're NOT using, and you may optionally rename
the .efi file for the binary you are using. If this is an initial
installation, you should rename refind.conf-sample to refind.conf; but if
you're replacing an existing installation, you should leave your existing
refind.conf intact. The end result might include the following files on the
ESP:

 EFI/refind/refind_x64.efi
 EFI/refind/refind.conf
 EFI/refind/icons/

Unfortunately, dropping the files in the ESP is not sufficient; as
described in the docs/installing.html file, you must also tell your EFI
about rEFInd. Precisely how to do this varies with your OS or, if you
choose to do it through the EFI, your EFI implementation. In some cases you
may need to rename the EFI/refind directory as EFI/boot, and rename
refind_x86.efi to bootx64.efi (or refind_ia32.efi to bootia32.efi on 32-bit
systems). Consult the installing.html file for full details.

Brief Installation Instructions (Source Package)
================================================

rEFInd source code can be obtained from
https://sourceforge.net/projects/refind/. Consult the BUILDING.txt file in
the source code package for build instructions. Once  you've built the
source code, you should duplicate the directory tree described above by
copying the individual files and the icons directory to the ESP. Note that
the binary file created by the build process will be called "refind.efi".
You can use that name or rename it to include your architecture code, as
you see fit.


NEWS.txt
0.3.5 (5/15/2012):
------------------

- Removed the GRUB 2 detection "reciped" added with 0.3.2, since I've
  received reports that it's not working as intended.

- Added re-scan feature: Press the Esc key to have rEFInd re-read its 
  configuration file, tell the EFI to scan for new filesystems, and re-scan
  those filesystems for boot loaders. The main purpose is to enable
  scanning a new removable medium that you insert after launching rEFInd;
  however, it can also be used to immediately implement changes to the
  configuration file or new drivers you load from an EFI shell.

- Fixed a bug that could cause the scroll-right arrow to be replaced by the
  scroll-left arrow under some circumstances.

0.3.4 (5/9/2012):
-----------------

- Added new configuration file option: "icons_dir", which sets the name
  of the subdirectory in which icons are found. See the documentation or
  sample configuration file for a full description.

- Modified Makefile to generate rEFInd binary that includes architecture
  code -- refind_ia32.efi or refind_x64.efi, rather than the generic
  refind.efi. This is done mainly to help the install.sh script. The
  program can be named anything you like on the disk. (The generic name
  refind.efi is used on unknown architectures.)

- Improved install.sh script: Fixed bug on OS X 10.7 and enable it to be
  used after building from source code (or via new "make install" Makefile
  target).

- Improved screen redraws to produce less flicker when moving among the
  second-row tags or to the last tag on the first row.

0.3.3 (5/6/2012):
-----------------

- Improved menu navigation:
  - In graphics mode, left & right arrow keys move left & right, while up &
    down arrows move between rows.
  - Page Up and Page Down now move through chunks of visible tags (in both
    text & graphics modes), jumping from one row to another only when at
    the edge of the row. In text mode, the "rows" are broken down as in
    graphics mode, but they aren't visibly distinguished on the screen.

- Improved text-mode use: rEFInd now displays the proper number of entries
  when first started in text mode and scrolling is done sensibly when too
  many entries exist to fit on the screen.

0.3.2 (5/4/2012):
-----------------

- Added the install.sh script to install rEFInd on Linux and Mac OS X
  systems. This script must be run as root (or via sudo). It requires
  no options, but on Mac OS X, passing it the "esp" option causes it
  to install rEFInd on the computer's ESP rather than the default of the
  currently OS X boot partition. (Under Linux, the default is to install to
  the ESP.) Note that there may be some unusual cases in which this script
  will fail to work.

- Does a better job of clearing the screen when launching OSes in text
  mode.

- Added detection "recipe" for GRUB 2's BIOS Boot Partition.

- Fixed bogus detection of ESPs created by Linux's mkdosfs utility or
  Windows as  bootable partitions when "scanfor" includes BIOS scanning
  options.


0.3.1 (4/27/2012):
------------------

- Fixed bug that caused spurious "Unsupported while scanning the root
  directory" messages under some conitions on Macs.

- Modified loader scanning code to sort boot loader entries within a
  directory by modification time, so that the most recently-modified loader
  is first among those in a given directory. Thus, if you specify a
  directory name (or volume name, for loaders stored in the root directory
  of a volume) as the default_selection, the most recent of those loaders
  will be the default. This is intended to help with Linux kernel
  maintenance when using the EFI stub loader; set up this way, the most
  recent kernel copied to your kernel directory will be the default,
  obviating the need to adjust the refind.conf file when adding a new
  kernel. If you want to change the default among those in the default
  directory, you can use "touch" to adjust the modification timestamp.

- Tweaked code to find loader-specific .icns file so that it finds files
  for Linux kernels without .efi extensions. In this case, files should be
  named the same as the kernels they match, but with .icns extensions. For
  instance, bzImage-3.3.2 should have an icon called bzImage-3.3.2.icns.
  (The old code would have looked for an icon called bzImage-3.3.icns.)

- Eliminated bogus OS loader tags for filenames that end in ".icns" when
  the scan_all_linux_kernels option is set.

0.3.0 (4/22/2012):
------------------

- I'm officially upgrading this project's status from "alpha" to "beta" and
  giving it a bump from 0.2.x to 0.3.0. This doesn't reflect any major
  milestone with this version; rather, it reflects my sense that rEFInd has
  been "out there" for a while, and although I've gotten bug reports,
  they've been minor and/or have been fixed. The program still has known
  bugs, but my impression is that it is, overall, usable by ordinary users.

- Added "resolution" option to refind.conf, which enables setting the video
  resolution. To use it, pass two numeric values, as in "resolution 1024
  768" to use a 1024x768 video mode. Note that not all modes are supported.
  If you specify a non-supported video mode on a UEFI system, a message
  appears listing the supported video modes and you must then press a key
  to continue, using the default video mode (usually 800x600).
  Unfortunately, I don't know the calls to get a list of supported video
  modes on older EFI 1.x systems (including Macs), so on Macs setting an
  incorrect video mode silently fails (you keep using the default mode).
  This makes changing your video mode a hit-or-miss proposition on Macs.
  CAUTION: It's possible to set a legal video mode that your monitor can't
  handle, in which case you'll get a blank display until you boot an OS
  that resets the video mode.

- Fixed (maybe) a bug that caused rEFInd to crash when returning from an
  EFI shell or other programs on Macs, particularly when rEFInd used

install.sh
#!/bin/bash
#
# Linux/MacOS X script to install rEFInd
#
# Usage:
#
# ./install.sh [esp]
#
# The "esp" option is valid only on Mac OS X; it causes
# installation to the EFI System Partition (ESP) rather than
# to the current OS X boot partition. Under Linux, this script
# installs to the ESP by default.
#
# This program is copyright (c) 2012 by Roderick W. Smith
# It is released under the terms of the GNU GPL, version 3,
# a copy of which should be included in the file COPYING.txt.
#
# Revision history:
#
# 0.3.3.1 -- Fixed OS X 10.7 bug; also works as make target
# 0.3.2.1 -- Check for presence of source files; aborts if not present
# 0.3.2   -- Initial version
#
# Note: install.sh version numbers match those of the rEFInd package
# with which they first appeared.

TargetDir=/EFI/refind

#
# Functions used by both OS X and Linux....
#

# Abort if the rEFInd files can't be found.
# Also sets $ConfFile to point to the configuration file, and
# $IconsDir to point to the icons directory
CheckForFiles() {
   # Note: This check is satisfied if EITHER the 32- or the 64-bit version
   # is found, even on the wrong platform. This is because the platform
   # hasn't yet been determined. This could obviously be improved, but it
   # would mean restructuring lots more code....
   if [[ ! -f $RefindDir/refind_ia32.efi && ! -f $RefindDir/refind_x64.efi ]] ; then
      echo "The rEFInd binary file is missing! Aborting installation!"
      exit 1
   fi

   if [[ -f $RefindDir/refind.conf-sample ]] ; then
      ConfFile=$RefindDir/refind.conf-sample
   elif [[ -f $ThisDir/refind.conf-sample ]] ; then
      ConfFile=$ThisDir/refind.conf-sample
   else
      echo "The sample configuration file is missing! Aborting installation!"
      exit 1
   fi

   if [[ -d $RefindDir/icons ]] ; then
      IconsDir=$RefindDir/icons
   elif [[ -d $ThisDir/icons ]] ; then
      IconsDir=$ThisDir/icons
   else
      echo "The icons directory is missing! Aborting installation!"
   fi
} # CheckForFiles()

# Copy the rEFInd files to the ESP or OS X root partition.
# Sets Problems=1 if any critical commands fail.
CopyRefindFiles() {
   mkdir -p $InstallPart/$TargetDir &> /dev/null
   if [[ $Platform == 'EFI32' ]] ; then
      cp $RefindDir/refind_ia32.efi $InstallPart/$TargetDir
      if [[ $? != 0 ]] ; then
         Problems=1
      fi
      Refind="refind_ia32.efi"
   elif [[ $Platform == 'EFI64' ]] ; then
      cp $RefindDir/refind_x64.efi $InstallPart/$TargetDir
      if [[ $? != 0 ]] ; then
         Problems=1
      fi
      Refind="refind_x64.efi"
   else
      echo "Unknown platform! Aborting!"
      exit 1
   fi
   echo "Copied rEFInd binary file $Refind"
   echo ""
   if [[ -d $InstallPart/$TargetDir/icons ]] ; then
      rm -rf $InstallPart/$TargetDir/icons-backup &> /dev/null
      mv -f $InstallPart/$TargetDir/icons $InstallPart/$TargetDir/icons-backup
      echo "Notice: Backed up existing icons directory as icons-backup."
   fi
   cp -r $IconsDir $InstallPart/$TargetDir
   if [[ $? != 0 ]] ; then
      Problems=1
   fi
   if [[ -f $InstallPart/$TargetDir/refind.conf ]] ; then
      echo "Existing refind.conf file found; copying sample file as refind.conf-sample"
      echo "to avoid collision."
      echo ""
      cp -f $ConfFile $InstallPart/$TargetDir
      if [[ $? != 0 ]] ; then
         Problems=1
      fi
   else
      echo "Copying sample configuration file as refind.conf; edit this file to configure"
      echo "rEFInd."
      echo ""
      cp -f $ConfFile $InstallPart/$TargetDir/refind.conf
      if [[ $? != 0 ]] ; then
         Problems=1
      fi
   fi
} # CopyRefindFiles()


#
# A series of OS X support functions....
#

# Mount the ESP at /Volumes/ESP or determine its current mount
# point.
# Sets InstallPart to the ESP mount point
# Sets UnmountEsp if we mounted it
MountOSXESP() {
   # Identify the ESP. Note: This returns the FIRST ESP found;
   # if the system has multiple disks, this could be wrong!
   Temp=`diskutil list | grep EFI`
   Esp=/dev/`echo $Temp | cut -f 5 -d ' '`
   # If the ESP is mounted, use its current mount point....
   Temp=`df | grep $Esp`
   InstallPart=`echo $Temp | cut -f 6 -d ' '`
   if [[ $InstallPart == '' ]] ; then
      mkdir /Volumes/ESP &> /dev/null
      mount -t msdos $Esp /Volumes/ESP
      if [[ $? != 0 ]] ; then
         echo "Unable to mount ESP! Aborting!\n"
         exit 1
      fi
      UnmountEsp=1
      InstallPart="/Volumes/ESP"
   fi
} # MountOSXESP()

# Control the OS X installation.
# Sets Problems=1 if problems found during the installation.
InstallOnOSX() {
   echo "Installing rEFInd on OS X...."
   if [[ $1 == 'esp' || $1 == 'ESP' ]] ; then
      MountOSXESP
   else
      InstallPart="/"
   fi
   echo "Installing rEFInd to the partition mounted at '$InstallPart'"
   Platform=`ioreg -l -p IODeviceTree | grep firmware-abi | cut -d "\"" -f 4`
   CopyRefindFiles
   if [[ $1 == 'esp' || $1 == 'ESP' ]] ; then
      bless --mount $InstallPart --setBoot --file $InstallPart/$TargetDir/$Refind
   else
      bless --setBoot --folder $InstallPart/$TargetDir --file $InstallPart/$TargetDir/$Refind
   fi
   if [[ $? != 0 ]] ; then
      Problems=1
   fi
   echo
   echo "WARNING: If you have an Advanced Format disk, *DO NOT* attempt to check the"
   echo "bless status with 'bless --info', since this is known to cause disk corruption"
   echo "on some systems!!"
   echo
} # InstallOnOSX()


#
# Now a series of Linux support functions....
#

# Identifies the ESP's location (/boot or /boot/efi); aborts if
# the ESP isn't mounted at either location.
# Sets InstallPart to the ESP mount point.
FindLinuxESP() {
   EspLine=`df /boot/efi | grep boot`
   InstallPart=`echo $EspLine | cut -d " " -f 6`
   EspFilesystem=`grep $InstallPart /etc/mtab | cut -d " " -f 3`
   if [[ $EspFilesystem != 'vfat' ]] ; then
      echo "/boot/efi doesn't seem to be on a VFAT filesystem. The ESP must be mounted at"
      echo "/boot or /boot/efi and it must be VFAT! Aborting!"
      exit 1
   fi
   echo "ESP was found at $InstallPart using $EspFilesystem"
} # MountLinuxESP

# Uses efibootmgr to add an entry for rEFInd to the EFI's NVRAM.
# If this fails, sets Problems=1
AddBootEntry() {
   Efibootmgr=`which efibootmgr 2> /dev/null`
   if [[ $Efibootmgr ]] ; then
      modprobe efivars &> /dev/null
      InstallDisk=`grep $InstallPart /etc/mtab | cut -d " " -f 1 | cut -c 1-8`
      PartNum=`grep $InstallPart /etc/mtab | cut -d " " -f 1 | cut -c 9-10`
      EntryFilename=$TargetDir/$Refind
      EfiEntryFilename=`echo ${EntryFilename//\//\\\}`
      ExistingEntry=`$Efibootmgr -v | grep $Refind`
      if [[ $ExistingEntry ]] ; then
         echo "An existing EFI boot manager entry for rEFInd seems to exist:"
         echo
         echo "$ExistingEntry"
         echo
         echo "This entry is NOT being modified, and no new entry is being created."
      else
         $Efibootmgr -c -l $EfiEntryFilename -L rEFInd -d $InstallDisk -p $PartNum &> /dev/null
         if [[ $? != 0 ]] ; then
	    EfibootmgrProblems=1
            Problems=1
         fi
      fi
   else
      EfibootmgrProblems=1
      Problems=1
   fi
   if [[ $EfibootmgrProblems ]] ; then
      echo
      echo "ALERT: There were problems running the efibootmgr program! You may need to"
      echo "rename the $Refind binary to the default name (EFI/boot/bootx64.efi"
      echo "on x86-64 systems or EFI/boot/bootia32.efi on x86 systems) to have it run!"
      echo
   fi
} # AddBootEntry()

# Controls rEFInd installation under Linux.
# Sets Problems=1 if something goes wrong.
InstallOnLinux() {
   echo "Installing rEFInd on Linux...."
   FindLinuxESP
   CpuType=`uname -m`
   if [[ $CpuType == 'x86_64' ]] ; then
      Platform="EFI64"
   elif [[ $CpuType == 'i386' || $CpuType == 'i486' || $CpuType == 'i586' || $CpuType == 'i686' ]] ; then
      Platform="EFI32"
      echo
      echo "CAUTION: This Linux installation uses a 32-bit kernel. 32-bit EFI-based"
      echo "computers are VERY RARE. If you've installed a 32-bit version of Linux"
      echo "on a 64-bit computer, you should manually install the 64-bit version of"
      echo "rEFInd. If you're installing on a Mac, you should do so from OS X. If"
      echo "you're positive you want to continue with this installation, answer 'Y'"
      echo "to the following question..."
      echo
      echo -n "Are you sure you want to continue (Y/N)? "
      read ContYN
      if [[ $ContYN == "Y" || $ContYN == "y" ]] ; then
         echo "OK; continuing with the installation..."
      else
         exit 0
      fi
   else
      echo "Unknown CPU type '$CpuType'; aborting!"
      exit 1
   fi
   CopyRefindFiles
   AddBootEntry
} # InstallOnLinux()

#
# The main part of the script. Sets a few environment variables,
# performs a few startup checks, and then calls functions to
# install under OS X or Linux, depending on the detected platform.
#

OSName=`uname -s`
ThisDir="$( cd -P "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
RefindDir="$ThisDir/refind"
ThisScript="$ThisDir/`basename $0`"
CheckForFiles
if [[ `whoami` != "root" ]] ; then
   echo "Not running as root; attempting to elevate privileges via sudo...."
   sudo $ThisScript $1
   if [[ $? != 0 ]] ; then
      echo "This script must be run as root (or using sudo). Exiting!"
      exit 1
   else
      exit 0
   fi
fi
if [[ $OSName == 'Darwin' ]] ; then
   InstallOnOSX $1
elif [[ $OSName == 'Linux' ]] ; then
   InstallOnLinux
else
   echo "Running on unknown OS; aborting!"
fi

if [[ $Problems ]] ; then
   echo
   echo "ALERT:"
   echo "Installation has completed, but problems were detected. Review the output for"
   echo "error messages and take corrective measures as necessary. You may need to"
   echo "re-run this script or install manually before rEFInd will work."
   echo
else
   echo
   echo "Installation has completed successfully."
   echo
fi

if [[ $UnmountEsp ]] ; then
   umount $InstallPart
fi
